Fix Wamouracampa stance switching

Skerxan opened the pull request:
-- They wouldn't switch their stance at all.
-- Used !mobsub command to determine the ID's for their stances.
-- 44 is stretched out, 45 is curled up.
-- BLU should finally be able to learn Cannonball.

<!-- remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm: -->
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] that I agree to LandSandBoat's [Limited Contributor License Agreement](https://github.com/LandSandBoat/server/blob/base/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I have **read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)**
- [x] that I've _**tested my code and things my code changed**_ since the last commit in the PR, and will test after any later commits

Skerxan committed:
Fix Wamouracampa stance switching

-- The wouldn't switch their stance at all.
-- Used !mobsub command to determine the ID's for their stances.
-- 44 is stretched out, 45 is curled up.
-- BLU should finally be able to learn Cannonball.

Skerxan commented:
#791

zach2good commented:
I don't understand what's going on here:

The logic here is:
```
on spawn:
formTime localVar is set to unix time + random 43-45 seconds

on roam and on fight:
if you're in animationSub 0 (standing) and unix time is larger than formTime: change animationSub to 1 and update formTime
if you're in animationSub 1 (standing) and unix time is larger than formTime: change animationSub to 0 and update formTime
```

Your changes only alter the randomness added to formTime from `math.random(43, 47)`  to `math.random(44, 45)`

zach2good commented:
As I look at it, the logic seems reasonable, but `formTime` must be being set incorrectly in the first place - so the condition where ` os.time()` is larger than it never happens

zach2good commented:
Maybe you pasted into the wrong place, it seems like the animationSub of 4/5 changes their form:
```lua
-----------------------------------
-- Area: Halvung
--  Mob: Wamouracampa
-- TODO: Damage resistances in streched and curled stances. Halting movement during stance change. Morph into Wamoura.
-----------------------------------
require("scripts/globals/status")
-----------------------------------
local entity = {}

entity.onMobSpawn = function(mob)
    mob:getLocalVar("formTime", math.random(0, 60)) -- Spread randomness
end

local tryChangeForm = function(mob)
    local formTime = mob:getLocalVar("formTime")
    local diff = os.time() - formTime
    local rand = math.random(43, 47)

    if diff > 0 then
        if mob:getAnimationSub() == 4 then
            mob:setAnimationSub(5)
            mob:setLocalVar("formTime", os.time() + rand)
        elseif mob:getAnimationSub() == 5 then
            mob:setAnimationSub(4)
            mob:setLocalVar("formTime", os.time() + rand)
        end
    end
end

entity.onMobRoam = function(mob)
    tryChangeForm(mob)
end

entity.onMobFight = function(mob, target)
    tryChangeForm(mob)
end

entity.onMobDeath = function(mob)
end

return entity
```

zach2good commented:
If you make these changes I'd happily accept them, but if you fancy a challenge, this sort of functionality would work much better as a family mixin, you can see how they work with `Uragnites`:

https://github.com/LandSandBoat/server/blob/base/scripts/mixins/families/uragnite.lua

```lua
-----------------------------------
-- Area: Bibiki Bay
--  Mob: Coralline Uragnite
-----------------------------------
mixins = {require("scripts/mixins/families/uragnite")}
-----------------------------------
local entity = {}

entity.onMobDeath = function(mob, player, isKiller)
end

return entity
```

Skerxan commented:
Thanks alot for all the input. Sorry for the trouble. Trying to fix it with your and Xaver's help. Bare with me xD;

Skerxan committed:
Complete rework with family mixin

Bluemages of LSB rejoice! Cannonball is otw.

Huge shoutout to @Xaver-DaRed for his coding prowess and working together with me on this issue for a whole day.

Wamouracampa behavior has been mostly accurate captured directly from retail. Not every feature is implemented like Wamoura transformation but for that I atleast uploaded a video in my mentioned #issue above.

Skerxan committed:
Fixed errors of commit

Xaver-DaRed commented:
I cannot review this in good faith, but I must say it has been tested and it works pretty well.

It was compared against retail for the HP margins and timers. While they may not be a 1:1 its pretty close, IMO.

zach2good commented on sql/mob_skill_lists.sql:
```diff
@@ -1130,9 +1130,10 @@ INSERT INTO `mob_skill_lists` VALUES ('Wamoura',253,1955);
--- INSERT INTO `mob_skill_lists` VALUES ('Wamouracampa',254,1820);
+INSERT INTO `mob_skill_lists` VALUES ('Wamouracampa',254,1819); -- Heat Barrier
+INSERT INTO `mob_skill_lists` VALUES ('Wamouracampa',254,1820);
+INSERT INTO `mob_skill_lists` VALUES ('Wamouracampa',1162,1818); -- Cannonball
```
Please put these new entries at the bottom of the file. I know it looks janky to have things split, but it acts as another safety check for us to not reuse IDs (a huge problem in the past).

Also, the names here are just syntax sugar, so you can rename them to be Wamouracampa_Standing or Wamouracampa_Ball etc.

Add comments for every entry etc.

zach2good commented on scripts/mixins/families/wamouracampa.lua:
```diff
@@ -0,0 +1,101 @@
+    mob:setAnimationSub(5) -- Curl
+    mob:setLocalVar("formTimeRoam", os.time() + math.random(43, 47))
+    mob:setMobMod(xi.mobMod.SKILL_LIST, 1162) -- Set Curled Skill List. ("Cannonball" and "Heat Barrier" only)
+    -- TODO: Change resistances.
```
I think it would be OK to add placeholder resistances here. I didn't check both wikis, but one did mention that one form is DEF and the other is MDEF.

You could cache the starting DEF/MDEF on spawn, and then use +/- 30% of that for each stat and leave a comment saying that this is a guess.

Xaver-DaRed replied:
After some testing in retail by Skerxan, we concluded that:
While streched, Phisical Damage Taken = 100%, Magic Damage Taken = 75%
While Curled, Phisical Damage Taken = 75%, Magic Damage Taken = 100%

The testing done, however, isn't enough to be considered conclusive. It's a guess, but a close one.

zach2good reviewed (commented):
Just a little cleanup, then should be good to go 👍

Xaver-DaRed commented:
Found this...
https://wiki-ffo-jp.translate.goog/html/9888.html?_x_tr_sch=http&_x_tr_sl=ja&_x_tr_tl=en&_x_tr_hl=es&_x_tr_pto=sc

It states that "ball" form has enhanced defence.

Skerxan committed:
Added def/mdef mods to stances

Skerxan commented:
Should be ready to be reviewed now.
